### DISCUSSION TITLE
[@types/eslint] `rules` option not typed

### DISCUSSION THREAD

**Discussion by reverofevil** [DefinitelyTyped/DefinitelyTyped#53901] [Issues with a @types package]
I've made a [feature request](https://github.com/microsoft/TypeScript/issues/44636) to TypeScript to add `typeof import()` by literal string type. Currently it's awaiting feedback, and I thought it would be wise to ask authors of typings it might mostly affect to give it.

As I understand, `eslint` lacks typing for rules, because rules get imported dynamically, and there is no way to put types there without this new language feature. Here's how it could look like:

```ts
// imported-rule1.d.ts
declare module 'imported-rule1' {
    export const create: (context: RuleContext<[foo: number]>) => {};
}

// imported-rule2.d.ts
declare module 'imported-rule2' {
    export const create: (context: RuleContext<[bar: 'option1' | 'option2']>) => {};
}

// eslint-lib.d.ts
type RuleContext<Options extends any[]> = {
    options: Options;
    // ...
}
type RuleListener = Record<string, (...args: any[]) => void> // stub

interface RuleModule<Options extends any[]> {
    create(context: RuleContext<Options>): RuleListener;
}

// hacks
type _FakeTypeofImportMap = {
    'imported-rule1': typeof import('imported-rule1');
    'imported-rule2': typeof import('imported-rule2');
    'not-a-rule': {},
}
type _FakeTypeofImport<ModuleName extends string> =
    ModuleName extends keyof _FakeTypeofImportMap
        ? _FakeTypeofImportMap[ModuleName]
        : never

// eslint.d.ts
// type OptionsByName<RuleName extends string, Exports = typeof import(RuleName)> =
type OptionsByName<RuleName extends string, Exports = _FakeTypeofImport<RuleName>> =
    Exports extends RuleModule<infer Options>
        ? Options
        : never

type Severity = 0 | 1 | 2;
type RuleLevel = Severity | "off" | "warn" | "error";
type RuleLevelAndOptions<RuleName extends string, Options = OptionsByName<RuleName>> =
    Options extends any[]
        ? [...Options, RuleLevel]
        : never;

type RuleEntry<RuleName extends string> = RuleLevel | RuleLevelAndOptions<RuleName>;

type RulesRecord<RuleNames extends string> = {
    [RuleName in RuleNames]: RuleEntry<RuleName>;
}

interface BaseConfig<RuleNames extends string> {
    rules: RulesRecord<RuleNames>;
}

interface Options<RuleNames extends string> {
    baseConfig: BaseConfig<RuleNames>;
}

declare class ESLint<RuleNames extends string> {
    constructor(options?: Options<RuleNames>);
    // ...
}

// eslint-runner.ts
new ESLint({
    baseConfig: {
        rules: {
            "imported-rule1": [5, 2],
            "imported-rule2": ['option1', 1],
            "smth-else": 2,
        }
    }
});
```

[Live version.](https://www.typescriptlang.org/play?#code/PTAEEsFsAcHsCcAuBTAJgWngVwDbIIwB0qhiAzgFCrIDGOAhvMqJLKrswORRxJqYd8nUAG8KoCaGQAPXolA1YAOzLyaTeigBcoABSKlKaYh0AlDgGFlRxAB4A2gDNYsHUqyQARsngBdAHwAlKAAvP6iAL4A3BQRFBQgEDAIKBjYeABMxKSU1HSMzKzseKDcyXxpHBnCYpJSsikKyqoKGtp6BjZmltYydvaejDqcsNCI4MpCoAA+paPjytUBwWGRMXEJYMhkOOCG6Lue2eQUiACe0MzmeFaGfbYA8mMTKvUoSqhkoPRKZ-YBoVE4kk8xeZB0TwWKhidUShHhsVOFyuHAAMuBVMglD5AaZaAhULZVPA9gBzAA0enhhEYpPB31+-xW4QAbrBwKhwolVFhPPE9ih4I56DQUXgALJsDiPZ7NN5Yz4Mv4A2qSdTITTIfS9YzdG46uyQsFBPXIdGY7HwdbxRIACxFAGtKOdLqAAPoAMXoDuQABVkbBHABJcqIcX0aCA1USMpyfjpAicHQu5CBpJyXSxlLxwScQIwyRZioCTJJ0AptM8FKZqvFhPVfPAmNKWCIdD0EvIMsiCLkxEp91en3+y6BkNyWyS4rIABy9EgzD6Cq+xLJ4RCTdAU44c4X8o+Xx9ZzTnu9foDwdD4egm7qAH5B2eR6nL3Jr-Zt3hd8hfLfJG5kBZHwbS2HYBWOZ1kVAI1mgAITOb9bGuWd50XYxl1AVclApUAAFEGiQL4QkfYcL3HFIkJ3VD-HXTd8LkL4lwPUBkM-ZBbD2RwcRglR-D-CQHx4yg6jqACgPgeIBwAZUAnxwHOQEAAYZlAfAVIyGIB2Q1FZJwQEZPE+SzhUgAiQNHBM0yAHdGCUSzZhMnx4AQEzNKg7TdIAQQ+ITKK-VD90VLCcKEwEhPgxDkO-GjQk3UKmMVH5lX40AH3sakhMpDygJwX8RP-UBsXEmIkVdZDcMMeAzj8lC9wSldEBJbD1xYtFdJU7LkBwbzUF8qLqJKrSODIPFFHgQl+oXRj0OY4KWujUB7Em5g9la-ypt8U0Ksa6rlv8a0KAFHxhVFUBYPoMhkFuRxwFJGrv2m94gsatcgTqBN6WQkb8XG+7ULIfbEUOu4hRFZg+qoqbAoaprSXCBbBku67bp0c6keUG67uWgGDryBgmAUBgyC+XCpPRQw-qh+rMJe5q3rVZpGqwGhEAQXRQWaO8IVlFRKe2IICwkOEEQ2RJtl2fZsCUS0cgobErLwsmBV0BGLqujGUfp-KPp0Bb8okEza1STt8BMnR7AAVkpDJfD7fWRMN0Mc0yM3FpGHmhEpfBbZSuoTLISBEFtdAusu12Mjt-W4jqOIInzIA)

Would this new language feature help typing `eslint`? Please, leave any feedback on the [mentioned issue](https://github.com/microsoft/TypeScript/issues/44636). Probably, it would make sense to keep this issue open until it gets implemented.

@pmdartus @j-f1 @saadq @JasonHK @bradzacher @JounQin

**Comment by bradzacher** [comment]
It's not typed because it is very hard to actually type it, and it would rely upon the user being able to specify all their plugin types ahead of time.

Someone would also have to maintain types for each plugin by manually converting JSON schemas to TS types.
This is a cumbersome process in of itself, and whilst there are tools that can help do this - it's cumbersome because some plugins declare bad JSON schemas.

A lot of plugins aren't written in TS - so it's unlikely that they will do it themselves. ESLint core is a good example of this.
The ESLint types are already sparsely maintained without including the maintenance overhead of rule schemas!

I think there's value in a typed config ecosystem - but I don't think doing it via DefinitelyTyped is the correct way to do it.

**Reply by reverofevil** [reply]
I totally forgot to set the topic. I intended to ask "whether this feature would help typing it", not whether `eslint` should be correctly typed in first place. I updated the topic. Sorry for that.

Otherwise, here are my 2 cents.

>  it is very hard to actually type it

Yes, it is, as well as all other libraries written in JS.

> would rely upon the user being able to specify all their plugin types

The code I provided shows that the user doesn't have to specify plugin types, if there are typings available on DefinitelyTyped.

> Someone would also have to maintain types for each plugin by manually converting JSON schemas to TS types
> it's cumbersome because some plugins declare bad JSON schemas.

There's a lot that could be done by automatically converting them and pushing them to DefinitelyTyped, and if there are wrong schemas, they will be found in this way. I should mention that React went through this process with their PropTypes, and we finally got typed React infrastructure.

> A lot of plugins aren't written in TS - so it's unlikely that they will do it themselves.

That's what we have DefinitelyTyped for: for libraries that don't have the types themselves, and don't want to write them. Several thousand typings there prove that there actually are people that are ready to spend their time.

**Comment by bradzacher** [comment]
> There's a lot that could be done by automatically converting them and pushing them to DefinitelyTyped

And who is doing this "a lot that could be done"?
Who is maintaining it?

Which is why I was getting at in my response.
This is all achievable.. with a coordinated team of people to build and maintain the infrastructure.

> Several thousand typings there prove that there actually are people that are ready to spend their time.

There aren't that many people who want to contribute to the ESLint world. There are less still that want to contribute to the subset of ESLint types. There are even less still that would want to contribute to types for ESLint plugins.

> and if there are wrong schemas, they will be found in this way

Assuming the plugin authors are active and willing to merge a (breaking change) to "fix" their schema.
Which (from my experience) they aren't always either.

Defining the types by itself in an extensible way that allows all plugins to ambiently include their types into the one interface isn't especially easy.

---

Again - I'm not saying that it's not possible to do this --- It's all very much technically possible.

I just think that it's not easily maintainable, and ultimately does not provide a lot of benefit vs the very probable and very real risk that the types will fall quickly out of date with nobody updating them.

IMO loose types (`{[ruleId: string]: GenericRuleConfig}`) is better than having wrong types.
